bokehjs/src/coffee/renderer/glyph/bokehgl.coffee

```diff
@@ -0,0 +1,347 @@
+        gl_FragColor = outline(distance, v_linewidth, u_antialias, v_fg_color, v_bg_color);
+        //gl_FragColor.rgb *= gl_FragColor.a;  // pre-multiply alpha
+    }
+    """
+    
+  MARKERCODE: "<defined in subclasses>" 
+     
+  init: () ->
+     
+    gl = @gl
+    frag = @FRAG.replace /MARKERCODE/, @MARKERCODE
+    
+    @last_trans = {}  # Keep track of transform
+    
+    # The program
+    @prog = new gloo2.Program(gl)
+    @prog.set_shaders(@VERT, frag)
+    @vbo_x = new gloo2.VertexBuffer(gl)
+    # Real attributes
+    @prog.set_attribute('a_x', 'float', [@vbo_x, 0, 0])
+    @vbo_y = new gloo2.VertexBuffer(gl)
+    @prog.set_attribute('a_y', 'float', [@vbo_y, 0, 0])
+    @vbo_s = new gloo2.VertexBuffer(gl)
+    @prog.set_attribute('a_size', 'float', [@vbo_s, 0, 0])
+    @vbo_a = new gloo2.VertexBuffer(gl)
+    @prog.set_attribute('a_angle', 'float', [@vbo_a, 0, 0])
+    # VBO's for attributes (they may not be used if value is singleton)
+    @vbo_linewidth = new gloo2.VertexBuffer(gl)
+    @vbo_fg_color = new gloo2.VertexBuffer(gl)
+    @vbo_bg_color = new gloo2.VertexBuffer(gl)
+    @index_buffer = new gloo2.IndexBuffer(gl)
+ 
+  draw: (indices, mainGlyph, trans) ->
+    
+    # The main glyph has the data, *this* glyph has the visuals.
+    nvertices = mainGlyph.glglyph.nvertices
+    
+    # Upload data if we must. Only happens for main glyph.
+    if @data_changed
+      @_set_data(nvertices)
+      @data_changed = false
+    else if @glyph.radius? and (trans.sx != @last_trans.sx or trans.sy != @last_trans.sy)
+      # Keep screen radius up-to-date for circle glyph. Only happens when a radius is given
+      @last_trans = trans
+      @vbo_s.set_data(0, new Float32Array((s*2 for s in @glyph.sradius)))
+
+    # Update visuals if we must. Can happen for all glyphs.
+    if @visuals_changed
+      @_set_visuals(nvertices)
+      @visuals_changed = false
+    
+    # Handle transformation to device coordinates
+    @prog.set_uniform('u_canvas_size', 'vec2', [trans.width, trans.height])
+    @prog.set_uniform('u_offset', 'vec2', [trans.dx[0], trans.dy[0]])
+    @prog.set_uniform('u_scale', 'vec2', [trans.sx, trans.sy])
+    
+    # Select buffers from main glyph 
+    # (which may be this glyph but maybe not if this is a (non)selection glyph)
+    @prog.set_attribute('a_x', 'float', [mainGlyph.glglyph.vbo_x, 0, 0])
+    @prog.set_attribute('a_y', 'float', [mainGlyph.glglyph.vbo_y, 0, 0])
+    @prog.set_attribute('a_size', 'float', [mainGlyph.glglyph.vbo_s, 0, 0])
+    
+    # Draw directly or using indices. Do not handle indices if they do not 
+    # fit in a uint16; WebGL 1.0 does not support uint32.
+    if indices.length == 0
+      return
+    else if indices.length == nvertices
+      @prog.draw(@gl.POINTS, [0, nvertices])
+    else if nvertices > 65535
+      @prog.draw(@gl.POINTS, [0, nvertices])
+      console.log('BOKEHGL ERROR: cannot handle selections for datasets of over 65535 elements.')
```

Comment:
I'd have to look into it a bit better to see how feasible it is, but I think we can draw in chunks (multiple calls to `@prog.draw()`)